### PR TITLE
Refactor the approach for recursive init calls

### PR DESCRIPTION
**Pull request by levitte**
At any point, we can expect OPENSSL_init_crypto() to be called
recursively.  This works, as long as no locking happens in code that's
executed again in the recursion.  We already had a similar situation
with the ERR code, and solved it with a rather ugly special case.

Here, we change the approach by keeping track of what initializations
have already been started, and simply never try to start them again.
We do so by masking away accumulated previous opts from the current
ones in the OPENSSL_init_crypto call.

There's a consequence, that any init opt can only be used once, even
if that particular initialization ends with aan error.

### CODE REVIEW

File: crypto/err/err.c [hunk 1]
```diff
@@ -146,13 +147,51 @@ static int err_string_data_cmp(const ERR_STRING_DATA *a,
...
+
+DEFINE_RUN_ONCE_STATIC(do_err_local_init)
+{
+    set_err_thread_local = 1;
+    return CRYPTO_THREAD_init_local(&err_thread_local, NULL);
+}
+
+static int err_local_init(void)
+{
+    return OPENSSL_init_crypto(0, NULL)
```
Comment (mattcaswell) [inline crypto/err/err.c]: Why is this needed? My assumption is we need to init threads because of the thread local storage. There needs to be a comment against all of these explaining it. I think it is probably better to auto-initialise threading from inside the `CRYPTO_THREAD_init_local` call.

File: crypto/init.c [hunk 2]
```diff
@@ -491,43 +499,61 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
...
+     * weren't already there.
+     *
+     * Note that this really makes any init option a one chance only, even
+     * if one of the initializations fail.
      */
+    {
+        CRYPTO_THREAD_write_lock(init_lock);
+        opts &= ~last_opts;      /* Mask away what's already been done */
+        last_opts |= opts;       /* Add what we're currently going to do */
+        CRYPTO_THREAD_unlock(init_lock);
```
Comment (mattcaswell) [inline crypto/init.c]: This is going to do the wrong thing in multi-threaded code. Thread 1 gets to this code first and removes stuff from last_opts that its about to do and then gets de-scheduled before it actually initialises them. Thread 2 then gets initialised gets to here and sees that it doesn't need to initialise things and returns. It then continues expecting those things to have been initialised and fails.

File: crypto/store/store_init.c [hunk 3]
```diff
@@ -14,13 +14,13 @@
...
 {
-    return OPENSSL_init_crypto(0, NULL)
-        && ossl_store_file_loader_init();
+    return ossl_store_file_loader_init();
 }
 
 int ossl_store_init_once(void)
 {
-    if (!RUN_ONCE(&store_init, do_store_init)) {
+    if (!OPENSSL_init_crypto(0, NULL)
```
Comment (mattcaswell) [inline crypto/store/store_init.c]: Why?

File: crypto/bio/b_addr.c [hunk 4]
```diff
@@ -603,12 +603,16 @@ static int addrinfo_wrap(int family, int socktype,
...
 {
-    if (!OPENSSL_init_crypto(0, NULL))
-        return 0;
     bio_lookup_lock = CRYPTO_THREAD_lock_new();
     return bio_lookup_lock != NULL;
 }
 
+static int bio_lookup_init(void)
+{
+    return OPENSSL_init_crypto(0, NULL)
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/bio/b_addr.c]: Apparently I am to blame for this (135648bcd0). This needs a rethink.

**Comment by mattcaswell** [comment]
I think this can probably be closed now. This PR was a precursor to #9372 which was ultimately done in a different way via #9492.

**Comment by levitte** [comment]
Yeah, thanks for the reminder.  I'd like to revisit the process at some point, but that's currently fairly low on my list, and I need to start over anyway when I get there.
